source/vof/handler.cc

```diff
@@ -0,0 +1,337 @@
+    : fraction (fraction),
+      reconstruction (reconstruction),
+      level_set (level_set),
+      c_field_name (c_field_name)
+  {}
+
+  template <int dim>
+  VoFHandler<dim>::VoFHandler (Simulator<dim> &simulator,
+                               ParameterHandler &prm)
+    : sim (simulator),
+      vof_initial_conditions (VoFInitialConditions::create_initial_conditions<dim>(prm)),
+      assembler ()
+  {
+    this->initialize_simulator(sim);
+    assembler.initialize_simulator(sim);
+    parse_parameters (prm);
+    assembler.set_vof_epsilon(vof_epsilon);
+
+    sim.signals.edit_finite_element_variables.connect(std_cxx11::bind(&aspect::VoFHandler<dim>::edit_finite_element_variables,
+                                                                      std_cxx11::ref(*this),
+                                                                      std_cxx11::_1));
+    sim.signals.post_set_initial_state.connect(std_cxx11::bind(&aspect::VoFHandler<dim>::set_initial_vofs,
+                                                               std_cxx11::ref(*this)));
+  }
+
+  template <int dim>
+  void
+  VoFHandler<dim>::edit_finite_element_variables (std::vector<VariableDeclaration<dim> > &vars)
+  {
+    for (unsigned int f=0; f<n_vof_fields; ++f)
+      {
+        vars.push_back(VariableDeclaration<dim>("vof_"+vof_field_names[f],
+                                                std_cxx11::shared_ptr<FiniteElement<dim>>(
+                                                  new FE_DGQ<dim>(0)),
+                                                1,
+                                                1));
+
+        vars.push_back(VariableDeclaration<dim>("vofN_"+vof_field_names[f],
+                                                std_cxx11::shared_ptr<FiniteElement<dim>>(
+                                                  new FE_DGQ<dim>(0)),
+                                                dim+1,
+                                                1));
+
+        vars.push_back(VariableDeclaration<dim>("vofLS_"+vof_field_names[f],
+                                                std_cxx11::shared_ptr<FiniteElement<dim>>(
+                                                  new FE_DGQ<dim>(1)),
+                                                1,
+                                                1));
+      }
+  }
+
+  template <int dim>
+  void
+  VoFHandler<dim>::declare_parameters (ParameterHandler &prm)
+  {
+    prm.enter_subsection ("VoF config");
+    {
+      prm.declare_entry ("Number of fields", "1",
+                         Patterns::Integer(0),
+                         "The number of fields to be handled usingg VoF interface tracking.");
+
+      prm.declare_entry ("Small volume", "1e-6",
+                         Patterns::Double (0, 1),
+                         "Minimum significant volume. VOFs below this considered to be zero.");
+
+      prm.declare_entry ("VoF solver tolerance", "1e-12",
```

Comment:
I guess it would be better to have separate solver tolerances for each compositional field, then you could just use the `Composition solver tolerance` for the VoF fields instead of introducing this new parameter. But until that is possible, this will be ok.